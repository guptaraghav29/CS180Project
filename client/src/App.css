@tailwind base;
@tailwind components;
@tailwind utilities;

.App {
  text-align: center;
  display: flex;
  flex-direction: column;
  align-items: center;
}

html,
body {
  scroll-behavior: smooth;
  padding: 2em;
  margin: 2px;
  text-align: center;
  font-family: "Work Sans", sans-serif;
  font-size: 22px;
  /* overflow-y: scroll; */
}
