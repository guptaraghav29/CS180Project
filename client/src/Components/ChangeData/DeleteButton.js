import React from "react";
import Button from "@mui/material/Button";
import DeleteIcon from "@mui/icons-material/Delete";

const DeleteButton = ({ currentSelection, fetchData }) => {
  return (
    <div className="">
      <Button
        variant="contained"
        sx={{
          background: "#ff5552",
          ":hover": {
            background: "#ff3733",
          },
        }}
        startIcon={<DeleteIcon />}
        onClick={() => {
          currentSelection.forEach((id) => {
            fetch(`/cars/${id}`, { method: "DELETE" });
          });
          fetchData();
        }}
      >
        Delete
      </Button>
    </div>
  );
};

export default DeleteButton;
