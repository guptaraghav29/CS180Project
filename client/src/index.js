import React from "react";
import ReactDOM from "react-dom";
import "./index.css";
import App from "./App";
import { BrowserRouter, Routes, Route } from "react-router-dom";
import ChartTrends from "./Components/Analytics/ChartTrends";
import ImportCSV from "./Components/ChangeData/ImportCSV";


ReactDOM.render(
	<React.StrictMode>
		<BrowserRouter>
			<Routes>
				<Route path="/" element={<App />} />
				<Route path="ChartTrends" element={<ChartTrends />} />
				<Route path="ImportCSV" element={<ImportCSV/>} />
			</Routes>
		</BrowserRouter>
	</React.StrictMode>,
	document.getElementById("root")
);

// If you want to start measuring performance in your app, pass a function
// to log results (for example: reportWebVitals(console.log))
// or send to an analytics endpoint. Learn more: https://bit.ly/CRA-vitals
