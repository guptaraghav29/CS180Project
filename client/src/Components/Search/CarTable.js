import * as React from "react";
import { DataGrid } from "@mui/x-data-grid";
import DeleteButton from "../ChangeData/DeleteButton";
import AddButton from "../ChangeData/AddButton";
import SaveData from "../ChangeData/SaveData";
import AveragePrice from "../Analytics/AveragePrice";
import NumPostings from "../Analytics/NumPostings";
import BarChart from "../Graphs/BarChart";

const columns = [
  {
    field: "id",
    headerName: "ID",
    headerAlign: "center",
    flex: 1,
    editable: true,
    renderCell: (cellValues) => {
      return (
        <div
          label="tabel"
          style={{
            width: "100%",
            textAlign: "center",
          }}
        >
          {cellValues.value}
        </div>
      );
    },
  },
  {
    field: "brand",
    headerName: "Brand",
    headerAlign: "center",
    flex: 1,
    editable: true,
    renderCell: (cellValues) => {
      return (
        <div
          style={{
            width: "100%",
            textAlign: "center",
          }}
        >
          {cellValues.value}
        </div>
      );
    },
  },
  {
    field: "model",
    headerName: "Model",
    headerAlign: "center",
    flex: 1,
    editable: true,
    renderCell: (cellValues) => {
      return (
        <div
          style={{
            width: "100%",
            textAlign: "center",
          }}
        >
          {cellValues.value}
        </div>
      );
    },
  },
  {
    field: "year",
    headerName: "Year",
    headerAlign: "center",
    flex: 1,
    editable: true,
    renderCell: (cellValues) => {
      return (
        <div
          style={{
            width: "100%",
            textAlign: "center",
          }}
        >
          {cellValues.value}
        </div>
      );
    },
  },
  {
    field: "price",
    headerName: "Price",
    headerAlign: "center",
    flex: 1,
    editable: true,
    renderCell: (cellValues) => {
      return (
        <div
          style={{
            width: "100%",
            textAlign: "center",
          }}
        >
          ${cellValues.value}
        </div>
      );
    },
  },
  {
    field: "odometer",
    headerName: "Odometer",
    headerAlign: "center",
    flex: 1,
    editable: true,
    renderCell: (cellValues) => {
      return (
        <div
          style={{
            width: "100%",
            textAlign: "center",
          }}
        >
          {cellValues.value}
        </div>
      );
    },
  },
  {
    field: "title_status",
    headerName: "Status",
    headerAlign: "center",
    flex: 1,
    editable: true,
    renderCell: (cellValues) => {
      return (
        <div
          style={{
            width: "100%",
            textAlign: "center",
          }}
        >
          {cellValues.value}
        </div>
      );
    },
  },
  {
    field: "paint_color",
    headerName: "Color",
    headerAlign: "center",
    flex: 1,
    editable: true,
    renderCell: (cellValues) => {
      return (
        <div
          style={{
            width: "100%",
            textAlign: "center",
          }}
        >
          {cellValues.value}
        </div>
      );
    },
  },
  {
    field: "region",
    headerName: "Region",
    headerAlign: "center",
    flex: 1,
    editable: true,
    renderCell: (cellValues) => {
      return (
        <div
          style={{
            width: "100%",
            textAlign: "center",
          }}
        >
          {cellValues.value}
        </div>
      );
    },
  },
  {
    field: "state",
    headerName: "State",
    headerAlign: "center",
    flex: 1,
    editable: true,
    renderCell: (cellValues) => {
      return (
        <div
          style={{
            width: "100%",
            textAlign: "center",
          }}
        >
          {cellValues.value}
        </div>
      );
    },
  },
  {
    field: "posting_date",
    headerName: "Date Listed",
    headerAlign: "center",
    flex: 1,
    editable: true,
    renderCell: (cellValues) => {
      return (
        <div
          style={{
            width: "100%",
            textAlign: "center",
          }}
        >
          {cellValues.value}
        </div>
      );
    },
  },
];

const CarTable = ({
  carTableData,
  carCompanies,
  currentSelection,
  setCurrentSelection,
  fetchData,
  chartData,
}) => {
  const handleCellEdit = (event) => {
    fetch(`/cars/${event.id}/update`, {
      method: "POST",
      headers: {
        Accept: "application/json",
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        id: event.id,
        field: event.field,
        value: event.value,
      }),
    }).then((response) => {
      if (!response.status === 200) {
        console.log("error");
      }
    });
    fetchData();
  };

  return (
    <div
      style={{
        height: 500,
        width: "100%",
        background: "#f9f9f9",
        textAlign: "center",
      }}
    >
      <DataGrid
        rows={carTableData}
        columns={columns}
        pageSize={50}
        rowsPerPageOptions={[50]}
        checkboxSelection
        onCellEditCommit={handleCellEdit}
        sx={{
          boxShadow: 2,
          border: 0,

          "& .MuiDataGrid-row:hover": {
            background: "lightgreen",
          },
          "& .MuiDataGrid-row:nth-of-type(odd)": {
            background: "#EDEDED",
          },
          "& .MuiDataGrid-columnHeaderTitle": {
            fontWeight: "bold",
          },
          "& .MuiDataGrid-row:nth-of-type(odd):hover": {
            background: "lightgreen",
          },
        }}
        onSelectionModelChange={(ids) => {
          setCurrentSelection(ids);
        }}
      />
      <div className="py-3 flex justify-between">
        <div>
          <SaveData carTableData={carTableData} />
        </div>
        <div>
          <AveragePrice
            carTableData={carTableData}
            currentSelection={currentSelection}
          />
        </div>
        <div>
          <NumPostings
            carTableData={carTableData}
            currentSelection={currentSelection}
          />
        </div>
        <div className="flex">
          <AddButton carCompanies={carCompanies} fetchData={fetchData} />
          <DeleteButton
            currentSelection={currentSelection}
            setCurrentSelection={setCurrentSelection}
            fetchData={fetchData}
          />
        </div>
      </div>
      <div className="items-center px-8" style={{ width: 800 }}>
        <BarChart chartData={chartData} />
      </div>
    </div>
  );
};

export default CarTable;
