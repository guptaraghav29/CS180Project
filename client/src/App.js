import "./App.css";
import React, { useState, useEffect } from "react";
import CarBrands from "./Components/Search/CarBrands";
import CarModels from "./Components/Search/CarModels";
import CarYears from "./Components/Search/CarYears";
import CarTable from "./Components/Search/CarTable";
import NavBar from "./Components/NavBar/NavBar";

function App() {
  const [carData, setCarData] = useState([]);
  const [carCompanies, setCarCompanies] = useState([]);
  const [currentCompany, setCurrentCompany] = useState("");
  const [companyModels, setCompanyModels] = useState([]);
  const [currentModel, setCurrentModel] = useState("");
  const [modelYears, setModelYears] = useState([]);
  const [currentYear, setCurrentYear] = useState("");
  const [carTableData, setCarTableData] = useState([]);
  const [currentSelection, setCurrentSelection] = useState([]);

  useEffect(() => {
    fetchData();
  }, []);

  const fetchData = () => {
    fetch("/cars")
      .then((response) => response.json())
      .then((data) => {
        setCarData(JSON.parse(JSON.stringify(data)));
        let cars = [];
        data.forEach((car) => {
          if (
            cars.indexOf(car.manufacturer) === -1
          ) {
            cars.push(car.manufacturer);
          }
        });
        cars.sort();
        setCarCompanies(cars);
      })
      .catch((error) => console.log(error));
  };

  useEffect(() => {
    let carModels = [];
    carData.forEach((car) => {
      if (
        car.manufacturer === currentCompany &&
        carModels.indexOf(car.model) === -1 &&
        car.model.length > 0
      ) {
        carModels.push(car.model);
      }
    });
    carModels.sort();
    setCompanyModels(carModels);
  }, [currentCompany, carData]);

  useEffect(() => {
    let years = [];
    carData.forEach((car) => {
      if (car.model === currentModel && years.indexOf(car.year) === -1) {
        years.push(car.year);
      }
    });
    years.sort();
    setModelYears(years);
  }, [currentModel, carData]);

  useEffect(() => {
    let carTable = [];

    carData.forEach((car) => {
      if (
        car.manufacturer === currentCompany &&
        car.model === currentModel &&
        car.year === currentYear
      ) {
        carTable.push({
          id: car.id,
          region: car.region,
          brand: car.manufacturer,
          model: car.model,
          year: car.year,
          price: car.price,
          odometer: car.odometer,
          title_status: car.title_status,
          paint_color: car.paint_color,
          state: car.state.toUpperCase(),
          posting_date: car.posting_date,
        });
      }
    });
    setCarTableData(carTable);
  }, [carData, currentCompany, currentModel, currentYear]);

  return (
    <div className="flex flex-col items-center">
      <div id="models" className="py-8">
        <NavBar />
      </div>
      <h1 className="center text-4xl mb-6"> Used Car Dataset </h1>
      <div className="flex" href="models">
        <div className="py-2 px-2">
          <CarBrands
            carCompanies={carCompanies}
            currentCompany={currentCompany}
            setCurrentCompany={setCurrentCompany}
            setCurrentModel={setCurrentModel}
          />
        </div>
        <div className="py-2 px-2">
          <CarModels
            companyModels={companyModels}
            currentModel={currentModel}
            setCurrentModel={setCurrentModel}
            setCurrentYear={setCurrentYear}
            style={{
              display:
                currentCompany !== "" && currentCompany !== null
                  ? "block"
                  : "none",
            }}
          />
        </div>
        <div className="py-2 px-2">
          <CarYears
            modelYears={modelYears}
            currentYear={currentYear}
            setCurrentYear={setCurrentYear}
            style={{
              display:
                currentModel !== "" && currentModel !== null ? "block" : "none",
            }}
          />
        </div>
      </div>

      <div className="py-8 w-full">
        {currentCompany !== "" &&
        currentCompany !== null &&
        currentModel !== "" &&
        currentModel !== null &&
        currentYear !== "" &&
        currentYear !== null ? (
          <div className="px-8">
            <CarTable
              carTableData={carTableData}
              carCompanies={carCompanies}
              currentSelection={currentSelection}
              setCurrentSelection={setCurrentSelection}
              fetchData={fetchData}
            />
          </div>
        ) : (
          <div></div>
        )}
      </div>
    </div>
  );
}

export default App;
