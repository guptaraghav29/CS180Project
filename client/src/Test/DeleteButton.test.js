import * as React from "react";
 import * as ReactDOM from "react-dom";
import DeleteButton from "../Components/ChangeData/DeleteButton";
 import App from "../App";
import renderer from "react-test-renderer";
import { createRoot } from 'react-dom/client';
import CarTable from "../Components/Search/CarTable";

import {
    fireEvent,
    getQueriesForElement,
    render,
} from "@testing-library/react";


it("testing the snapshot", () => {
	const tree = renderer.create(< DeleteButton />).toJSON();
	expect(tree).toMatchSnapshot();
});

it("renders without any crashes", () => {
	const div = document.createElement("div");
	const root = createRoot(div);
	root.render(< DeleteButton />);
});

it("should have Delete message", () => {
    const root = document.createElement("root");
	ReactDOM.render(<DeleteButton/>, root);
    const { getByText } = getQueriesForElement(root);
    expect(getByText("Delete")).not.toBeNull();
})


