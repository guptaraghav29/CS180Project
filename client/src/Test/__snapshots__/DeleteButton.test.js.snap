// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing the snapshot 1`] = `
<div
  className=""
>
  <button
    className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-bhxy9h-MuiButtonBase-root-MuiButton-root"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="DeleteIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
        />
      </svg>
    </span>
    Delete
  </button>
</div>
`;
